src/v/cluster/controller_backend.cc

```diff
@@ -1554,31 +1554,44 @@ ss::future<std::error_code> controller_backend::shutdown_on_current_shard(
 ss::future<> controller_backend::delete_partition(
   model::ntp ntp, model::revision_id rev, partition_removal_mode mode) {
     auto part = _partition_manager.local().get(ntp);
-    if (unlikely(part.get() == nullptr)) {
-        return ss::make_ready_future<>();
+    auto is_partition_replicated_locally = part.get() != nullptr;
+
+    // The partition leaders table contains partition leaders for all
+    // partitions accross the cluster. For this reason we need to delete from
+    // it regardless of whether a replica of 'ntp' is present on the node, if
+    // the removal mode is global.
+    auto f = [this, ntp, rev, is_partition_replicated_locally, mode]() {
+        if (
+          mode == partition_removal_mode::global
+          || is_partition_replicated_locally) {
+            return _partition_leaders_table.invoke_on_all(
+              [ntp, rev](auto& leaders) { leaders.remove_leader(ntp, rev); });
+        } else {
+            return ss::make_ready_future<>();
+        }
+    };
+
+    if (unlikely(!is_partition_replicated_locally)) {
+        return f();
     }
 
     // partition was already recreated with greater rev, do nothing
     if (unlikely(part->get_revision_id() > rev)) {
-        return ss::make_ready_future<>();
+        return f();
     }
 
     auto group_id = part->group();
 
-    return _shard_table
-      .invoke_on_all([ntp, group_id, rev](shard_table& st) mutable {
-          st.erase(ntp, group_id, rev);
-      })
-      .then([this, ntp, rev] {
-          return _partition_leaders_table.invoke_on_all(
-            [ntp, rev](partition_leaders_table& leaders) {
-                leaders.remove_leader(ntp, rev);
-            });
-      })
-      .then([this, ntp = std::move(ntp), mode] {
-          // remove partition
-          return _partition_manager.local().remove(ntp, mode);
-      });
+    return f().then([this, ntp, group_id, rev, mode]() {
+        return _shard_table
+          .invoke_on_all([ntp, group_id, rev](shard_table& st) mutable {
+              st.erase(ntp, group_id, rev);
+          })
+          .then([this, ntp = ntp, mode] {
+              // remove partition
+              return _partition_manager.local().remove(ntp, mode);
+          });
+    });
```

Comment:
nit: I think coroutinizing this will make it a bit easier. As is, it's a but hard to follow with `is_partition_replicated_locally` being conditioned in a couple places. Maybe something like (excuse the lazy code):

```
if (mode == partition_removal_mode::global || is_partition_replicated_locally) {
  co_await ...partition_leaders_table.remove_leader(...);
}

if (!is_partition_replicated_locally || part->get_revision_id() > rev) {
    co_return;
}

auto group_id = part->group();
co_await ...shard_table.erase(...);
co_await ...partition_manager.remove(...);
```

Reply:
agree that it would be more readable with coroutines

Reply:
Thanks for the suggestion. Coroutinised it.